src/compose/service.coffee

```diff
@@ -0,0 +1,732 @@
+			@volumes.push('/lib/modules:/lib/modules')
+		if checkTruthy(@labels['io.resin.features.firmware']) and opts.hostPathExists.firmware
+			@volumes.push('/lib/firmware:/lib/firmware')
+		if checkTruthy(@labels['io.resin.features.balena-socket'])
+			@volumes.push('/var/run/balena.sock:/var/run/balena.sock')
+			@environment['DOCKER_HOST'] ?= 'unix:///var/run/balena.sock'
+		if checkTruthy(@labels['io.resin.features.supervisor-api'])
+			@_addSupervisorApi(opts)
+		else
+			# We ensure the user hasn't added "supervisor0" to the service's networks
+			delete @networks[constants.supervisorNetworkInterface]
+		if checkTruthy(@labels['io.resin.features.resin-api'])
+			@environment['RESIN_API_KEY'] = opts.deviceApiKey
+
+	extendEnvVars: ({ imageInfo, uuid, appName, name, version, deviceType, osVersion }) =>
+		newEnv =
+			RESIN_APP_ID: @appId.toString()
+			RESIN_APP_NAME: appName
+			RESIN_SERVICE_NAME: @serviceName
+			RESIN_DEVICE_UUID: uuid
+			RESIN_DEVICE_NAME_AT_INIT: name
+			RESIN_DEVICE_TYPE: deviceType
+			RESIN_HOST_OS_VERSION: osVersion
+			RESIN_SUPERVISOR_VERSION: version
+			RESIN_APP_LOCK_PATH: '/tmp/resin/resin-updates.lock'
+			RESIN_SERVICE_KILL_ME_PATH: '/tmp/resin/resin-kill-me'
+			RESIN: '1'
+			USER: 'root'
+		if @environment?
+			_.defaults(newEnv, @environment)
+		_.defaults(newEnv, conversions.envArrayToObject(imageInfo?.Config?.Env ? []))
+		@environment = newEnv
+		return @environment
+
+	extendLabels: (imageInfo) =>
+		@labels = _.clone(@labels)
+		_.defaults(@labels, imageInfo?.Config?.Labels ? {})
+		@labels['io.resin.supervised'] = 'true'
+		@labels['io.resin.app-id'] = @appId.toString()
+		@labels['io.resin.service-id'] = @serviceId.toString()
+		@labels['io.resin.service-name'] = @serviceName
+		return @labels
+
+	extendAndSanitiseExposedPorts: (imageInfo) =>
+		@expose = _.clone(@expose)
+		@expose = _.map(@expose, String)
+		if imageInfo?.Config?.ExposedPorts?
+			for own k, v of imageInfo.Config.ExposedPorts
+				port = k.match(/^([0-9]*)\/tcp$/)?[1]
+				if port? and !_.find(@expose, port)
+					@expose.push(port)
+
+		return @expose
+
+	extendAndSanitiseVolumes: (imageInfo) =>
+		volumes = []
+		for vol in @volumes
+			isBind = /:/.test(vol)
+			if isBind
+				[ bindSource, bindDest, mode ] = vol.split(':')
+				if !path.isAbsolute(bindSource)
+					# Rewrite named volumes to namespace by appId
+					volDefinition = "#{@appId}_#{bindSource}:#{bindDest}"
+					if mode?
+						volDefinition += ":#{mode}"
+					volumes.push(volDefinition)
+				else
+					console.log("Ignoring invalid bind mount #{vol}")
+			else
+				volumes.push(vol)
+		volumes = volumes.concat(@defaultBinds())
+		volumes = _.union(_.keys(imageInfo?.Config?.Volumes), volumes)
+		@volumes = volumes
+		return @volumes
+
+	getNamedVolumes: =>
+		defaults = @defaultBinds()
+		validVolumes = _.map @volumes, (vol) ->
+			if _.includes(defaults, vol) or !/:/.test(vol)
```

Comment:
It might be worth using `vol.includes(':')` rather than using the the regex engine.